xmppserver/src/main/java/org/jivesoftware/openfire/muc/MultiUserChatManager.java

```diff
@@ -991,4 +991,114 @@ public List<Multimap<String, String>> clusteringStateConsistencyReportForMucRoom
                 mucService.getServiceName()
             )).collect(Collectors.toList());
     }
+
+    /**
+     * Gets the number of MUC room retirees.
+     *
+     * @return the number of MUC room retirees.
+     */
+    public int getRetireeCount(String mucServiceName) {
+        Long serviceID = getMultiUserChatServiceID(mucServiceName);
+
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        int count = 0;
+        try {
+            con = DbConnectionManager.getConnection();
+            pstmt = con.prepareStatement(COUNT_RETIREE);
+            pstmt.setLong(1, serviceID);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
+                count = rs.getInt(1);
+            }
+        } catch (SQLException e) {
+            Log.error("An unexpected exception occurred while trying to count the number of MUC room retirees.", e);
+        } finally {
+            DbConnectionManager.closeConnection(rs, pstmt, con);
+        }
+        return count;
+    }
+
+
+    /**
+     * Gets the retirees for a MUC service.
+     *
+     * @param mucServiceName the name of the MUC service.
+     * @param startIndex the index of the first result to return.
+     * @param numResults the maximum number of results to return.
+     * @return a map of retirees.
+     */
+    public Collection<String> getRetirees(String mucServiceName, int startIndex, int numResults) {
+
+        Long serviceID = getMultiUserChatServiceID(mucServiceName);
+        List<String> retirees = new ArrayList<>();
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+
+        try {
+            con = DbConnectionManager.getConnection();
+
+            if ((startIndex == 0) && (numResults == Integer.MAX_VALUE)) {
+                pstmt = con.prepareStatement(GET_RETIREES);
+                pstmt.setLong(1, serviceID);
+                // Set the fetch size. This will prevent some JDBC drivers from trying
+                // to load the entire result set into memory.
+                DbConnectionManager.setFetchSize(pstmt, 500);
```

Comment:
I happened to run into this in an unrelated project: only setting the fetch size is not sufficient to prevent the entire result set from being loaded in memory (at least not for MSSQL). That requires a bit more configuration of the query/result set Please use the changes in https://github.com/igniterealtime/openfire-monitoring-plugin/pull/410 as a guide.

I'm assuming that you've copied this from another query somewhere in Openfire. If possible, could you modify that implementation with the same change please?

Reply:
I've updated all instances of this.